src/odemis/util/test/registration_test.py

```diff
@@ -0,0 +1,404 @@
+        for matrix in [
+            [(1, 0), (0, 1)],
+            [(0, 1), (-1, 0)],
+            [(-1, 0), (0, -1)],
+            [(0, -1), (1, 0)],
+            [(0, 1), (1, 0)],
+            [(-1, 0), (0, 1)],
+            [(0, -1), (-1, 0)],
+            [(1, 0), (0, -1)],
+        ]:
+            out = _canonical_matrix_form(matrix)
+            numpy.testing.assert_array_almost_equal(out, numpy.eye(2))
+
+    def test_random_input(self):
+        """
+        `_canonical_matrix_form()` should accept any 2x2 transformation matrix.
+        The resulting matrix should not contain a reflection (have a positive
+        determinant) and have a rotation between ±45°.
+
+        """
+        for i in range(50):
+            matrix = self._rng.standard_normal((2, 2))
+            with self.subTest(matrix=matrix, i=i):
+                matrix = _canonical_matrix_form(matrix)
+                tform = AffineTransform(matrix)
+                self.assertGreater(numpy.linalg.det(matrix), 0)
+                self.assertGreater(tform.rotation, -0.25 * numpy.pi)
+                self.assertLess(tform.rotation, 0.25 * numpy.pi)
+
+    def test_angle_difference(self):
+        """
+        When provided with an input transformation with positive determinant
+        (i.e. no reflection), the resulting change in rotation should be a
+        multiple of 90°.
+
+        """
+        for i in range(50):
+            matrix = self._rng.standard_normal((2, 2))
+            if numpy.linalg.det(matrix) < 0:
+                matrix = numpy.fliplr(matrix)
+            with self.subTest(matrix=matrix, i=i):
+                tform1 = AffineTransform(matrix)
+                tform2 = AffineTransform(_canonical_matrix_form(matrix))
+                delta = tform2.rotation - tform1.rotation
+                n = int(round(delta / (0.5 * numpy.pi)))
+                self.assertAlmostEqual(delta - n * 0.5 * numpy.pi, 0)
+
+
+class EstimateGridOrientationTest(unittest.TestCase):
+    """Unittest for `estimate_grid_orientation()`."""
+
+    def setUp(self):
+        """Ensure reproducible tests."""
+        self._rng = check_random_state(12345)
+
+    def test_full_grid(self):
+        """
+        `estimate_grid_orientation()` should return the expected
+        AffineTransform for a full grid.
+
+        """
+        for shape in [(5, 5), (8, 8)]:
+            for i in range(50):
+                with self.subTest(shape=shape, i=i):
+                    n, m = shape
+                    shuffle = self._rng.permutation(n * m)
+                    tform = random_affine_transform(self._rng)
+                    xy = tform.apply(unit_gridpoints(shape, mode="xy")[shuffle])
+                    out = estimate_grid_orientation(xy, shape, AffineTransform)
+                    for name in ("matrix", "translation"):
+                        numpy.testing.assert_array_almost_equal(
+                            getattr(tform, name), getattr(out, name)
+                        )
```

Comment:
This is hard to read. I would remove the for loop and write it like:
```
numpy.testing.assert_array_almost_equal(tform.matrix, out.matrix)
numpy.testing.assert_array_almost_equal(tform.translation, out.translation)
```

Reply:
done